golang/cosmos/types/address_hooks.go

```diff
@@ -18,48 +18,97 @@ type AddressRole string
 const (
 	RoleSender   AddressRole = "Sender"
 	RoleReceiver AddressRole = "Receiver"
+
+	AddressHookVersion     = 0
+	BaseAddressLengthBytes = 2
 )
 
-func trimSlashPrefix(s string) string {
-	return strings.TrimPrefix(s, "/")
+// AddressHookMagic is a magic byte prefix that identifies a hooked address.
+// Chosen to make bech32 address hooks that look like "agoric10rch..."
+var AddressHookMagic = []byte{0x78, 0xf1, 0x70 | AddressHookVersion}
+
+func init() {
+	if AddressHookVersion&0x0f != AddressHookVersion {
+		panic(fmt.Sprintf("AddressHookVersion must be less than 0x10, got 0x%x", AddressHookVersion))
+	}
 }
 
-// ExtractBaseAddress extracts the base address from a parameterized address.
-// It removes all subpath and query components from addr.
+// ExtractBaseAddress extracts the base address from an Address Hook.  It
+// returns addr verbatim if it is not an Address Hook.
 func ExtractBaseAddress(addr string) (string, error) {
-	parsed, err := url.Parse(addr)
+	baseAddr, _, err := SplitHookedAddress(addr)
 	if err != nil {
 		return "", err
 	}
+	return baseAddr, nil
+}
+
+// SplitHookedAddress splits a hooked address into its base address and hook data.
+// For the JS implementation, look at @agoric/cosmic-proto/src/address-hooks.js.
+func SplitHookedAddress(addr string) (string, []byte, error) {
+	prefix, payload, err := bech32.DecodeAndConvert(addr)
+	if err != nil {
+		return "", []byte{}, err
+	}
 
-	// Specify the fields and values we expect.  Unspecified fields will only
-	// match if they are zero values in order to be robust against extensions to
-	// the url.URL struct.
-	//
-	// Remove leading slashes from the path fields so that only parsed relative
-	// paths match the expected test.
-	expected := url.URL{
-		Path:        trimSlashPrefix(parsed.Path),
-		RawPath:     trimSlashPrefix(parsed.RawPath),
-		RawQuery:    parsed.RawQuery,
-		Fragment:    parsed.Fragment,
-		RawFragment: parsed.RawFragment,
+	bz := bytes.TrimPrefix(payload, AddressHookMagic)
+	if len(bz) == len(payload) {
+		// Return an unhooked address.
+		return addr, []byte{}, nil
+	}
```

Comment:
For forward compatibility, this should probably error upon encountering a hooked address with an unknown version.
```suggestion
	bz := bytes.TrimPrefix(payload, AddressHookMagic[:2])
	if len(bz) == len(payload) || len(bz) == 0 || bz[0] & 0x70 != 0x70 {
		// Return an unhooked address.
		return addr, []byte{}, nil
	}
	version := bz[0] & 0x0F
	bz = bz[1:]
	if version != AddressHookVersion {
		return "", []byte{}, fmt.Errorf("unsupported version %d", version)
	}
```
(and likewise in the JS version)

Reply:
@turadg , @LuqiPan ,

I suppose this shows that my assumption that the design was agreed by all the relevant parties is not correct.